src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,164 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Try to convert a serial of additions into a single multiplication. Also convert `(a * CON) + a` to `(CON + 1) * a` as
+// a side effect (see below). On success, a new MulNode is returned.
+//
+// We perform such conversions incrementally during iGVN by transforming left most nodes first and work up to the root
+// of the expression. In other words, we convert, at each iteration:
+//        a + a + a + ... + a
+//     => 2*a + a + ... + a
+//     => 3*a + ... + a
+//     => n*a
+//
+// Due to the iterative nature of iGVN, MulNode transformed from first few AddNode terms may be further transformed into
+// power-of-2 pattern. (e.g., 2 * a => a << 1, 3 * a => (a << 2) + a). We can't guarantee we'll always pick up
+// transformed power-of-2 patterns when term `a` is complex.
+//
+// Therefore, when matching the LHS `a * CON`, we match with best efforts by looking for the following patterns:
+//     - (1) Simple addition: a + a
+//     - (2) Simple lshift: a << CON
+//     - (3) Simple multiplication: CON * a
+//     - (4) Power-of-two addition: (a << CON1) + (a << CON2)
```

Comment:
```suggestion
//     - (1) Simple addition:       LHS = a + a
//     - (2) Simple lshift:         LHS = a << CON
//     - (3) Simple multiplication: LHS = CON * a
//     - (4) Power-of-two addition: LHS = (a << CON1) + (a << CON2)
```
I suggest adding the `LHS =` explicitly here. I tend to not always read all the explanatory text, and search for patterns first. Then it is not immediately clear that you are actually matching a `a + a + a` rather than a `a + a` for `1)`.